crates/turborepo-lib/src/task_hash.rs

```diff
@@ -0,0 +1,263 @@
+
+            // TODO: Look into making WorkspaceName take a Cow
+            let workspace_name = WorkspaceName::Other(task_id.package().to_string());
+
+            let package_file_hash_inputs = PackageFileHashInputs {
+                task_id: task_id.clone(),
+                task_definition,
+                workspace_name,
+            };
+
+            hash_tasks.push(package_file_hash_inputs);
+        }
+
+        let mut hashes = HashMap::with_capacity(hash_tasks.len());
+        let mut hash_objects = HashMap::with_capacity(hash_tasks.len());
+
+        for package_file_hash_inputs in hash_tasks {
+            let pkg = workspaces
+                .get(&package_file_hash_inputs.workspace_name)
+                .ok_or_else(|| {
+                    Error::MissingPackageJson(package_file_hash_inputs.workspace_name.to_string())
+                })?;
+
+            let package_path = pkg
+                .package_json_path
+                .parent()
+                .unwrap_or_else(|| AnchoredSystemPath::new("").unwrap());
+
+            let mut hash_object = scm.get_package_file_hashes(
+                &repo_root,
+                package_path,
+                &package_file_hash_inputs.task_definition.inputs,
+            )?;
+
+            if !package_file_hash_inputs.task_definition.dot_env.is_empty() {
+                let package_path = pkg
+                    .package_json_path
+                    .parent()
+                    .unwrap_or_else(|| AnchoredSystemPath::new("").unwrap());
+                let absolute_package_path = repo_root.resolve(package_path);
+                let dot_env_object = scm.hash_existing_of(
+                    &absolute_package_path,
+                    package_file_hash_inputs
+                        .task_definition
+                        .dot_env
+                        .iter()
+                        .map(|p| p.to_anchored_system_path_buf()),
+                )?;
+
+                for (key, value) in dot_env_object {
+                    hash_object.insert(key, value);
+                }
+            }
+
+            let file_hashes = FileHashes(hash_object);
+            let hash = file_hashes.clone().hash();
+
+            hashes.insert(package_file_hash_inputs.task_id.to_string(), hash);
+            hash_objects.insert(package_file_hash_inputs.task_id.to_string(), file_hashes);
+        }
+
+        Ok(PackageInputsHashes {
+            hashes: hashes,
+            expanded_hashes: hash_objects,
+        })
+    }
+}
+
+/// Caches package-inputs hashes, and package-task hashes.
+struct TaskHasher {
+    package_inputs_hashes: PackageInputsHashes,
+    package_task_env_vars: HashMap<TaskId<'static>, DetailedMap>,
+    package_task_hashes: HashMap<TaskId<'static>, String>,
+    package_task_framework: HashMap<TaskId<'static>, String>,
+    package_task_outputs: HashMap<TaskId<'static>, Vec<AnchoredSystemPathBuf>>,
```

Comment:
As part of this port can we also port this to be theadsafe like it is in Go with the RwLock? The tracker will need to be shared across tasks in Rust as well.